spark/src/main/scala/org/apache/hadoop/fs/versioned/VersionedFileSystem.scala

```diff
@@ -0,0 +1,132 @@
+package org.apache.hadoop.fs.versioned
+
+import java.net.URI
+import java.util.Objects
+
+import com.gu.tableversions.metastore.VersionPaths
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs._
+import org.apache.hadoop.fs.permission.FsPermission
+import org.apache.hadoop.util.Progressable
+
+class VersionedFileSystem extends FileSystem {
+
+  private var baseURI: URI = _
+  private var baseFS: FileSystem = _
+  private var mappings: Map[String, String] = _
```

Comment:
minor: 
maybe this could have a name that describes from what to what it maps ?
something like BaseToVersionedPaths or something like that